library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/impl/QuiltBlockInternals.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.block.extensions.impl;
+
+import org.quiltmc.qsl.block.extensions.api.data.BlockDataKey;
+import org.quiltmc.qsl.block.extensions.api.data.ExtraBlockData;
+import org.quiltmc.qsl.block.extensions.mixin.AbstractBlockAccessor;
+import net.minecraft.block.Block;
+import java.util.Map;
+
+public final class QuiltBlockInternals {
+	private QuiltBlockInternals() { }
+
+	public static ExtraBlockData computeExtraData(Block block) {
+		Block.Settings settings = ((AbstractBlockAccessor) block).getSettings();
+		BlockSettingsInternals internals = (BlockSettingsInternals) settings;
+
+		ExtraBlockData extraData = internals.qsl$getExtraData();
+		if (extraData == null) {
+			ExtraBlockData.Builder builder = ExtraBlockData.builder();
+			Map<BlockDataKey<?>, Object> map = internals.qsl$getSettingsMap();
+			if (map != null) {
+				copyFromMap(builder, map);
+			}
+			ExtraBlockData.OnBuild.EVENT.invoker().append(block, settings, builder);
+			internals.qsl$setExtraData(extraData = builder.build());
```

Comment:
Couldn't you give the QuiltBlockSettings extra data map directly to the builder, instead of copying everything to the builder?

Reply:
I don't think I understood this. This is pretty much the most direct way to get the data map into the builder while not exposing a way to create an invalid `ExtraBlockData` collection (I.E keys mapped to values with incompatible types).

Reply:
How would an invalid `ExtraBlockData` be created here?

Reply:
Exposing a constructor that allows passing the map into the data collection/builder directly could allow this to happen:
```java
BlockDataKey<Integer> a = new BlockDataKey<>(Integer.class, new Identifier("badmod", "hehe"));
HashMap<BlockDataKey<?>, Object> map = new HashMap<>();
map.put(a, "This is not an integer!");
ExtraBlockData.Builder builder = ExtraBlockData.builder(map); // this builder will now construct an invalid collection!
```

Reply:
Couldn't the constructor be an implementation detail? Yes, it may cause issues if it's public API ~~you could just cast the BlockDataKey anyway~~, but I'm saying it should be used here specifically.